harvester/core/tasks/elastic_search.py

```diff
@@ -0,0 +1,30 @@
+from datetime import datetime
+from itertools import chain
+
+from django.db.transaction import atomic, DatabaseError
+from django.utils.timezone import make_aware
+from celery import current_app as app
+
+from datagrowth.utils.iterators import ibatch
+
+from core.models import ElasticIndex, DatasetVersion
+
+
+@app.task(name="sync_indices")
+def sync_indices():
+    dataset_version = DatasetVersion.objects.get_latest_version()
+    indices_queryset = ElasticIndex.objects.filter(dataset_version=dataset_version)
+    try:
+        with atomic():
+            current_time = make_aware(datetime.now())
+            for index in indices_queryset.select_for_update(nowait=True):
+                documents_queryset = dataset_version.document_set.filter(
+                    modified_at__gte=index.pushed_at
+                )
+                for doc_batch in ibatch(documents_queryset, batch_size=32):
```

Comment:
Zouden we niet alle documents in een losse task kunnen verwerken? Zodat als er 1 mocht falen de rest wel verwerkt wordt? En vervolgens als die tasks verwerkt zijn de index pushen?

Reply:
Ik snap niet helemaal wat je bedoeld. Ik schrijf maar wat dingen hier beneden, misschien dat het dan duidelijker wordt. Anders even bellen :)

Deze task pushed alleen naar de index en verwerkt niets. We doen het in een batch, omdat index.push het in een batch doet en het 1 request wordt. Dat scheelt weer HTTP latency.

Bovendien zit er allerlei overhead op Celery tasks en Django, dus tenzij het long-running is kun je beter meerdere dingen per task uitvoeren. Het werkt daarin heel anders dan een Node (het is geen eventloop). Eventueel kunnen we er een eventloop van maken (dat zou een epic worden), maar batches hebben als voordeel dat je vaak maar 1x een database/service hoeft aan te spreken. Wel zonde om non-blocking te gaan schrijven om veel IOWait te introduceren. Ik vraag me af of dat het waard is.

Als er een push faalt, dan willen we niets aan index updaten, zodat het nog een keer geprobeerd wordt tijdens de volgende run. Dit runt elke 30 seconden.

Reply:
Ah ok, dan lijkt het mij zo beter. Ik begreep dan niet helemaal wat het deed